### PR TITLE
Improve status to and from `Box<dyn Error>` detection

### PR DESCRIPTION
**Pull request by seanmonstar**
Sorry, this also includes some trait alias cleanup that was causing conflicts when I was trying to update linkerd2.

### CODE REVIEW

File: src/status.rs [hunk 1]
```diff
@@ -253,84 +386,114 @@ impl Code {
...
+        let orig = Status::new(Code::Cancelled, "stop eet!");
+        let err = orig.to_h2_error();
+
+        assert_eq!(err.reason(), Some(h2::Reason::CANCEL));
+    }
+
+    #[test]
+    fn code_from_i32() {
+        // This for loop should catch if we ever add a new variant and don't
+        // update From<i32>.
```
Comment (hawkw) [inline src/status.rs]: 👍 smart!

File: src/body.rs [hunk 2]
```diff
@@ -54,42 +55,35 @@ where
 {}
 
 /// Dynamic `Send` body object.
-pub struct BoxBody<T = BytesBuf, E = ::Status> {
-    inner: Box<Body<Item = T, Error = E> + Send>,
+pub struct BoxBody {
```
[the remark this answers is not in the document]
Comment (seanmonstar) [reply src/body.rs]: I couldn't understand why, but there was some trait mismatching when trying to update the proxy, pointing at the item type. Once I killed that here, it fixed it there. I figured using the generic here is probably not common anyways.

File: src/client/client_streaming.rs [hunk 3]
```diff
@@ -33,9 +33,9 @@ impl<T, U, B: Body> ResponseFuture<T, U, B> {
 impl<T, U, B> Future for ResponseFuture<T, U, B>
 where T: Message + Default,
       U: Future<Item = Response<B>>,
-      U::Error: Into<Box<dyn std::error::Error>>,
+      U::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
```
[the remark this answers is not in the document]
Comment (seanmonstar) [reply src/client/client_streaming.rs]: Would be nice :)

File: src/generic/codec.rs [hunk 4]
```diff
@@ -182,59 +184,87 @@ where T: Encoder<Item = U::Item>,
...
     fn is_end_stream(&self) -> bool {
         false
     }
 
     fn poll_buf(&mut self) -> Poll<Option<Self::Item>, Status> {
-        match self.inner {
+        match self.inner.poll_encode(&mut self.buf) {
+            Ok(ok) => Ok(ok),
+            Err(status) => {
+                match self.role {
```
[the remark this answers is not in the document]
Comment (seanmonstar) [reply src/generic/codec.rs]: I opted for a match so as to require an exhaustive check.

File: src/body.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (hawkw) [reply src/body.rs]: Might be worth making sure nobody else is using this (@per-gron?) before merging, but you're right that it doesn't seem particularly common.

**Pull request merged**